haskell-src/Concordium/Types/ProtocolLevelTokens/CBOR.hs

```diff
@@ -623,6 +624,192 @@ newtype TokenHolderTransaction = TokenHolderTransaction
 decodeTokenHolderTransaction :: Decoder s TokenHolderTransaction
 decodeTokenHolderTransaction = TokenHolderTransaction <$> decodeSequence decodeTokenHolderOperation
 
+-- | Parse a 'TokenHolderTransaction' from a 'LBS.ByteString'. The entire bytestring must
+--  be consumed in the parsing.
+tokenHolderTransactionFromBytes :: LBS.ByteString -> Either String TokenHolderTransaction
+tokenHolderTransactionFromBytes lbs =
+    case CBOR.deserialiseFromBytes decodeTokenHolderTransaction lbs of
+        Left e -> Left (show e)
+        Right ("", res) -> return res
+        Right (remaining, _) ->
+            Left $
+                show (LBS.length remaining)
+                    ++ " bytes remaining after parsing token-holder transaction"
+
 -- | Encode a 'TokenHolderTransaction' as CBOR.
 encodeTokenHolderTransaction :: TokenHolderTransaction -> Encoding
 encodeTokenHolderTransaction = encodeSequence encodeTokenHolderOperation . tokenHolderTransactions
+
+-- | CBOR-encode a 'TokenHolderTransaction' to a (strict) 'BS.ByteString'.
+tokenHolderTransactionToBytes :: TokenHolderTransaction -> BS.ByteString
+tokenHolderTransactionToBytes = CBOR.toStrictByteString . encodeTokenHolderTransaction
+
+-- * Reject reasons
+
+-- | Details provided by the token module in the event of rejecting a transaction.
+data EncodedTokenRejectReason = EncodedTokenRejectReason
+    { -- | The type of the reject reason. At most 255 bytes.
+      etrrType :: !BSS.ShortByteString,
+      -- | (Optional) CBOR-encoded details.
+      etrrDetails :: !(Maybe BSS.ShortByteString)
+    }
+    deriving (Eq, Show)
+
+-- | Reasons that a token-holder operation might be rejected by the Token Module.
+data TokenHolderFailure
+    = -- | The recipient address was not valid.
+      RecipientNotFound
+        { -- | The index in the list of operations of the failing operation.
+          thfTransactionIndex :: !Word64,
```

Comment:
Since `transactionIndex` already exists in the blockchain world which could cause confusion if we also call this index `transactionIndex`.

```suggestion
          thfOperationIndex :: !Word64,
```